### DISCUSSION TITLE
About Denoflare's R2 `listDirectories`

### DISCUSSION THREAD

**Discussion by galpt** [skymethod/denoflare#27] [General]
About Denoflare's R2 `listDirectories`, is it possible to limit the dir listed to only the ones we want them to be "public" instead of listing the entire bucket?

Examples as shown below:

### Scenario 1
We could enable `listDirectories` at the **root level**, but make a separate optional config in the `~/.denoflare.jsonc` to only show certain directories that we want to be publicly listed. Anything else that aren't declared there should be private and will give a 404 response, unless the requested url path is correctly pointed to a certain file.

The `listDirectories` at the root level is useful to access custom `index.html` and/or `404.html` at the root level, so that `404.html` applies everytime someone gets the `404 status`.

### Scenario 2
If Scenario 1 is quite hard to achieve, Scenario 2 might be easier.
Here in Scenario 2, instead of enabling `listDirectories` at the **root level**, there should be an optional config in the `~/.denoflare.jsonc` to set a custom url path that we want to be listed as the root directory. So we will put the `404.html` to this custom root path.

It might not give the same level of privacy as Scenario 1, but it will still be useful in some way, knowing that we can make certain directory listed as public and use that as a root, instead of listing the entire bucket.
So we don't need to migrate files in that bucket, that we don't want to be listed, to a separate bucket.

### Additional Scenario
Either Scenario 1 or 2, both should have the ability to add multiple directories that we want to be listed, and keep anything else private.
This might be possible by declaring something in the `~/.denoflare.jsonc` like
```
"bindings": {
                "bucket": { "bucketName": "my-bucket" },
                "flags": { "value": "disallowRobots,emulatePages,listDirectories" },
                "allowIps": { "value": "1.2.3.4" },
                "directoryListingLimit": { "value": "20" },
                "allowCorsOrigins": { "value": "*" },
            },
"customDomains": [ "my-bucket.my-domain.com" ],
"customListDirectories": {
                "root": "<BucketName>/<NowWeAreInTheBucketRootLevel>/DirNameAsTheNewRoot",
                "notFound": "<BucketName>/<NowWeAreInTheBucketRootLevel>/DirNameAsTheNewRoot/404.html",
            },
```
The above config example might be suitable for Scenario 2, while for Scenario 1, listing the bucket at the root level isn't a good idea unless it's only for the custom `404.html` we want to put at the root level.

In the end, Scenario 1 will force denoflare to support multiple separate directories that we only want to be listed as public, while keeping anything else unlisted. Something like this
```
"bindings": {
                "bucket": { "bucketName": "my-bucket" },
                "flags": { "value": "disallowRobots,emulatePages,listDirectories" },
                "allowIps": { "value": "1.2.3.4" },
                "directoryListingLimit": { "value": "20" },
                "allowCorsOrigins": { "value": "*" },
            },
"customDomains": [ "my-bucket.my-domain.com" ],
"customListDirectories": {
                "root": "<BucketName>/<NowWeAreInTheBucketRootLevel>",
                "notFound": "<BucketName>/404.html",
                "dirYouWantToBePubliclyListedFromTheRootLevel": [ "public1", "public3", "public5" ]
            },
```

Maybe the only difference is that Scenario 2 only supports 1 custom "root dir" instead of using the entire bucket as the root,
while Scenario 1 kinda supports multiple "sub-root dir"  (the root Parent dir is the bucket with the custom `404.html` there, and the sub-root directories act as the ones you want to be publicly listed, and give `404.html` as a response whenever a user requests any other paths other than the ones listed in `"dirYouWantToBePubliclyListedFromTheRootLevel"` — unless the user correctly specifies a path that points to a certain file in the bucket).

---
I couldn't find a related example to do that from the [r2-public-read example](https://denoflare.dev/examples/r2-public-read).
I'd like to discuss it here first, whether the above scenarios are possible or not.

**Comment by johnspurlock-skymethod** [comment]
interesting - perhaps an optional metafile in each directory could support this in a simple way instead of tunneling yet more info through string params? this way you could declare even very complex access rules at many levels

Something like `_access` could be consulted and used, if present, and never returned itself (kind of like `_headers`)

**Reply by galpt** [reply]
Is this `_access` metafile already supported by denoflare now? If yes, I might miss something from the docs.
But yeah I guess using a file that is able to limit `listDirectories` from listing the entire bucket is really a good idea.

Or if we couldn't prevent denoflare from listing the entire bucket, we could at least prevent users from accessing any other files, that we don't want users to be able to access, that are listed by denoflare.
This way, it should give some privacy to the entire bucket in some way, even when files are listed.

We don't have to only specify each file path that's allowed to be accessed.
But only need a way so that users can only access files inside a certain directory that are listed in the bucket, that we are completely sure any files in that directory are okay to be accessed publicly.

Any files and/or directories outside that `allowedAccess` dir should give a response like `not allowed.` or something like that.

**Reply by johnspurlock-skymethod** [reply]
Right, it doesn't exist yet, but it's something I'd be willing to add to the Denoflare public read worker.

Maybe something like a `.gitignore` file, called `.r2ignore` that you could specify files or patterns you want to exclude
